Disconnect event handling for Linux and Macos

TirelessDev opened the pull request:
This PR extends the connectHandler behaviour to work as expected for MacOS and Linux implementations. 

The expected behaviour of `adapter.SetConnectHandler` is that the callback will be fired when peripherals are both connected **and** disconnected. This behaviour is crucial for the correct handling of device connection states such as in a BLE gateway. However, the current implementation for this on MacOS and Linux is that the callback is only fired on a connection, and when doing so the address information is not included. Also, from what I can tell, no functionality exists on the windows side of things but I have no capability to test any improvements at this stage. 

This PR improves the behaviour for connection event callbacks of `adapter.SetConnectHandler` by adding the device address to the function call. It also implements the functionality to detect disconnection events and subsequently trigger the callback, also with the address of the disconnected device.

On MacOS this functionality is implemented by simply implementing `DidDisconnectPeripheral`  signature on the `centralManagerDelegate` as exposed by `github.com/tinygo-org/cbgo`.

On Linux, this functionality is implemented by creating a watcher for `Device1` property changes on the `"Connected"` property. Both connection and disconnection events are handled this way. The way this is implemented could be extended to other property changes but more plumbing would need to be added to achieve this and at the moment I have not distilled the need to do so.

A minor change this PR brings is to apply the `params.ConnectionTimeout` parameter to the connection timeout behaviour on MacOS and to correctly cancel the connection if this timeout is triggered.

This PR does not change the API in any way but should provide some useful & expected functionality. Further work will need to be done to bring the Windows implementation into line. I have tested it on an M1 MacBook Pro and a Raspberry Pi 4.

Looking forward to hearing your suggestions for any changes/improvements.

TirelessDev commented on adapter_darwin.go:
```diff
@@ -100,6 +101,18 @@ func (cmd *centralManagerDelegate) DidDiscoverPeripheral(cmgr cbgo.CentralManage
+// DidDisconnectPeripheral when peripheral is disconnected.
+func (cmd *centralManagerDelegate) DidDisconnectPeripheral(cmgr cbgo.CentralManager, prph cbgo.Peripheral, err error) {
+	if err != nil {
+		fmt.Println(err.Error())
```
I am not sure of the correct way to handle this error here. Ideally, the callback would be able to accept it so we can surface it to the user in a meaningful way.

aykevl replied:
From the documentation:

> The error parameter contains the reason for the disconnection, unless the disconnect resulted from a call to [cancelPeripheralConnection(_:)](https://developer.apple.com/documentation/corebluetooth/cbcentralmanager/1518952-cancelperipheralconnection).

So it contains the reason for the disconnection (things like "timeout" I guess).

Printing the disconnect reason seems like a bad idea. Instead, either provide it to the package user (not currently possible with the API) or just ignore it: the important information is the fact that the device disconnected. Why it was disconnected, is less important.

TirelessDev replied:
Sweet, okay. Happy to just ignore it for now then. 

It would be good to have some mechanism in the future to expose it, but as the availability and type of the error are going to be platform dependent, it doesn't make it simple. Something to explore in a later PR.

TirelessDev commented on gap.go:
```diff
@@ -11,6 +11,9 @@ var (
 )
 
+// default connection timeout
+const DEFAULT_CONNECTION_TIMEOUT time.Duration = 10 * time.Second
```
This 10 second const came from the hardcoded value in the macos adapter connect implementation. I have surfaced it in the general `gap.go` function so it can be used by other implementations if required later.

aykevl replied:
10 seconds may be the default for macOS, but it is not the default everywhere.
I agree it would be useful to expose, but:

  * The name should follow Go naming convention, which is CamelCase.
  * It would have to actually match the default timeout. For example, it is 4 seconds on Nordic SoftDevices:

https://github.com/tinygo-org/bluetooth/blob/e79ea1e4e99c2628e7219de0bd88da59adaf3d60/gap_nrf528xx-central.go#L134

My suggestion would be to rename it, and either make this a platform specific constant or use the same constant everywhere. (The latter would be ideal, but is perhaps better done in a separate PR).

TirelessDev replied:
Good points. I have moved it back to `gap_darwin.go` and renamed it to `defaultConnectionTimeout`.

TirelessDev commented on gap_darwin.go:
```diff
@@ -127,10 +132,11 @@ func (a *Adapter) Connect(address Addresser, params ConnectionParams) (*Device,
 		p.SetDelegate(d.delegate)
 
-		a.connectHandler(nil, true)
+		a.connectHandler(adr, true)
```
I am not sure why the `adr` was not included in this to begin with?

TirelessDev commented on gap_darwin.go:
```diff
@@ -127,10 +132,11 @@ func (a *Adapter) Connect(address Addresser, params ConnectionParams) (*Device,
 		return d, nil
-	case <-time.NewTimer(10 * time.Second).C:
+	case <-time.NewTimer(timeout).C:
+		a.cm.CancelConnect(prphs[0]) // we need to cancel the connection if we have timed out ourselves
```
Connections can take longer than the default 10 seconds to establish in some cases. This makes sure that when the user is indicated a connection timeout, the connection doesn't actually continue to be established in the background.

aykevl replied:
Good idea.
It _might_ be a good idea to wait until the device is disconnected (`didDisconnectPeripheral`):
https://developer.apple.com/documentation/corebluetooth/cbcentralmanager/1518952-cancelperipheralconnection

TirelessDev replied:
Okay, so I have modified the `DidDisconnectPeripheral` handler to also interact with the `adapter.connectMap` in the same way as the `DidConnectPeripheral` handler. This way, the `prphCh` in the `adapter.Connect` function will get a response on connect and disconnect, and we can check the peripheral state when it arrives. 

This does mean that the switch case has to be wrapped so that when the timer expires, we continue waiting for this disconnection event to trigger. This poses a concern if nothing ever shows up, and I wonder if we should start some form of bail-out timer just in case.

TirelessDev commented on gap_linux.go:
```diff
@@ -284,25 +291,73 @@ func (a *Adapter) Connect(address Addresser, params ConnectionParams) (*Device,
+				switch changed.Name {
+				case "Connected":
+					// Send off a notification indicating we have connected or disconnected
+					d.adapter.connectHandler(d.address, d.device.Properties.Connected)
```
The connected state also comes through in the `propchanged` event under `changed.Value`. However, it seemed more appropriate to use the actual property it is notifying us of changes about. Either way works fine in my testing.

aykevl commented on gap_linux.go:
```diff
@@ -283,25 +290,73 @@ func (a *Adapter) Connect(address Address, params ConnectionParams) (*Device, er
+				}
+
+				continue
+			case <-d.ctx.Done():
+				d.propchanged <- nil
+				return
```
Why are you sending nil on this channel?

TirelessDev replied:
I honestly do not know. Probably a 2 am brain glitch. Have removed

aykevl reviewed (commented):
Thanks for working on this! This looks reasonable, with some comments.

TirelessDev committed:
macos: implement disconnect handeling

macos: add disconnect handler

macos: add default timeout and params.ConnectionTimeout handling


macos: add central manager CancelConnect call on connection timeout


macos: added address to connectHanndler call upon successful connection


macos: remove printing error on peripheral disconnection


macos: move and rename default connection timeout


macos: wait for disconnection upon timed out connection attempt

aykevl reviewed (approved):
Looks good to me.

Can you rebase these changes against the dev branch? There is a conflict. Also, it would be nice if you could squash some of the commits (for example, you could put all the Linux changes in one commit and all the macOS changes in another).

TirelessDev committed:
linux: implement disconnect handeling

linux: remove unnecessary sending of nil prop change  

linux: first pass at watching connection events


linux: remove unneeded debug prints


linux: fix spelling

linux: fix pointer address references

TirelessDev commented:
Okay, everything should be squashed and rebased. I have diffed against the previous version of the local branch, and everything looks identical.

aykevl commented:
Thanks a lot!

TirelessDev commented:
No problem, glad to help
